Use SlotNo and TimeInterpreter

Anviking opened the pull request:
# Issue Number

ADP-356 #1868 #1869 


# Overview

- [x] Add TimeInterpreter, necessary queries, and a way to run them
Add tests comparing them with old implementation

- [x] Fundamentally change wallet to use SlotNo instead of SlotId
    - Patch together cardano-wallet-jormungandr to still work, despite the
binary using SlotId.

- [x] Remove redundant byron code

- [x] Sync progress calculation compares times now instead of slots

- [x] Rename the `slot` tables to `slot_no` to force automatic migration, preventing catastrophic re-interpretation of old data

# Comments

- Depends on #1890 

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Convert wallet to use SlotNo and TimeInterpreter

== Add TimeInterpreter, necessary queries, and a way to run them
Add tests comparing them with old implementation

== Fundamentally change wallet to use SlotNo instead of SlotId
- Patch together cardano-wallet-jormungandr to still work, despite the
binary using SlotId.

== Remove redundant byron code

== Sync progress calculation compares times now instead of slots

Anviking commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -799,9 +798,11 @@ prop_listRegisteredPools DBLayer {..} entries =
 
     prop = do
+        -- TODO#1901: Is this change ok? Previously only (SlotInEpoch 0) were
+        -- generated.
```
@jonathanknowles do you know off the top of your head if there was a reason for only `SlotId`s with slotInEpoch 0 were generated?
(you recently edited here)

Anviking replied:
It seems to work 🤷‍♂️

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -66,31 +78,100 @@ import Numeric.Natural
+-- We cannot manually specify when the fetching happens.
+--
+-- This may or may not be what we actually want.
+type TimeInterpreter m = forall a. Qry a -> m a
```
☝️ We may now have a `TimeInterpreter IO` that _sometimes_ takes a long time to run a conversion query (when it needs to refetch the summary over LSQ).

Not sure if we should change…

The re-fetching should happen on very rare occasions though…

rvl replied:
I think it would be better to have the network layer query the Interpreter on a schedule, and stash it in a mutable variable. So that all queries can be pure.

Anviking replied:
>  So that all queries can be pure

I presume it would still be in `IO` (to access the var), so no change to the interface.

In that case I'd be concerned with what callers would see if there is nothing in the var. There are many call-sites, and I think blocking until it is available is the simplest thing to. Which is somewhat like this.

But preemptively refreshing _before_ a conversion fails, sounds sane to do… optimally.

rvl replied:
Let's see. I was thinking that the initial Interpreter will be one constructed from genesis parameters.
It will get updated once the node has synced far enough. (Question - how far is enough?)
This would prevent the wallet from being unresponsive while the node is syncing.

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -90,7 +92,7 @@ TxMeta
     txMetaStatus       W.TxStatus   sql=status
     txMetaDirection    W.Direction  sql=direction
-    txMetaSlot         W.SlotId     sql=slot
+    txMetaSlot         SlotNo       sql=slot
```
I suspect we may need manual migration

rvl replied:
Don't fret over DB migration if it's too hard. We can start from a fresh state directory if necessary.

Anviking replied:
On second thought, I fear/wonder if users would silently get corrupted data, as the meaning of the `slot` table has changed, but not the name.

Is there a good way to test / detect this?

_Or I guess testing manually using CLI works. I tried using the Daedalus mainnet installation, and replacing the wallet and node manually, but I think too many other things have changed for it to work._

rvl replied:
If the type of the column changes, I think persistent auto migrations will either throw an exception or helpfully delete the table.

Let's decide next week whether to automatically migrate or make a fresh state directory.

Anviking committed:
Re-write toApiWalletDelegation for new slotting API

Anviking committed:
Remove genSlotId weed

Anviking committed:
Remove redundant `GenState s` constraint

Anviking committed:
Fix mistake in DB property

It would group slots like:

Epoch 1:
  SlotNo 0
  SlotNo 8
Epoch 0:
  SlotNo 11

when it zipped the grouped slots together with the epochs.

I'm not sure how it worked previously and stopped working now.

I also added a somewhat ugly shrinker for the stake pool fixtures.

Anviking committed:
Rename sqlite `slot` fields to `slot_no` forcing migration

We previously encoded
    SlotId = (epoch, slotInEpoch)
in the `slot` tables as a single, flat integer (using an artificial
epoch length)

We are now storing the SlotNos.

It would be VERY BAD if a new wallet would read re-interpret the old
wallet data incorrectly.

Renaming the fields should trigger the automatic DB migration.

Anviking committed:
Revert "Rename sqlite `slot` fields to `slot_no` forcing migration"

This reverts commit 9eb843b6075d92f8a149479f9d8f9b4c5671d98e.

That commit broke a unit test failure for a manual migration. I am not
sure if the test could have been removed without consequences.

But weirdly, migration seems to work smoothly without it:

$ cardano-wallet-shelley-2020-07-06 serve --node-socket socket --testnet shelley_testnet-genesis.json --database wallet-db
$ # Create a wallet
$ cardano-wallet-shelley transaction list <WALLET ID>

$ cardano-wallet-shelley serve --node-socket socket --testnet shelley_testnet-genesis.json --database wallet-db
$ cardano-wallet-shelley transaction list <WALLET ID>

$ # The correct slots are still shown.

Looking in the DBs, it seems like wallet does detect that the schema has
changed, and re-syncs from genesis.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -59,54 +77,198 @@ import Data.Quantity
+singleEraInterpreter gp q = either bomb return $ runQuery q int
   where
-    k = fromIntegral $ getQuantity $ getEpochStability gp
+    bomb x = error $ "singleEraInterpreter: " <> show x
```
Could you add a comment saying that these queries will never fail?

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -20,6 +20,10 @@
 
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+-- Needed for (Buildable SlotNo)
+-- TODO#1901: How do we want to deal with the orphan (Buildable SlotNo) instance?
```
I would add an `Cardano.Orphans` module or something like that, where it would be acceptable to have `-fno-warn-orphans`.

Anviking replied:
👍  Done

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -12,11 +16,20 @@
+    , Qry
 
-      -- * Old functions
+      -- * Legacy api
```
Will these all be removed in a future PR, or moved to a tests-only module?

Anviking replied:
I suppose we should remove them completely, and re-write the tests to use the new ones.

Now we are testing the old ones, and that the new ones are equivalent to the old ones (Generators aren't perfect though).

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -559,15 +555,11 @@ mkWithdrawals proxy (ChimericAccount keyHash) amount
-defaultTTL epochLength slot =
-    (toSlotNo epochLength slot) + 7200
+defaultTTL ::  SlotNo -> SlotNo
+defaultTTL slot = slot + 7200
```
I like this. :-)

rvl reviewed (approved):
Nice work - this is a really big PR.
I have checked Cardano.Wallet.Primitive and Cardano.Wallet.Shelley.
Is there anything else I should pay attention to?

Anviking committed:
Mention that singleEraInterpreter cannot fail

Anviking committed:
Move (Buildable SlotNo) to new Orhphans module

Anviking committed:
Remove out-of-date TODOs

Anviking committed:
Adjust generators for SlottingSpec properties
